### ISSUE TITLE
Problem with FI_ADDR_NOTAVAIL in C++

### ISSUE THREAD

**Issue by jsquyres** [ofiwg/libfabric#2161]
When trying to compile a C++ application that includes the use of `FI_ADDR_NOTAVAIL`, one gets this kind of compile error:

```
my_program.cc:249: error: ‘UINT64_MAX’ was not declared in this scope
```

It looks like `FI_ADDR_NOTAVAIL` (and `FI_ADDR_UNSPEC`) is defined in `fabric.h` as:

```c
#define FI_ADDR_UNSPEC          UINT64_MAX
#define FI_ADDR_NOTAVAIL        UINT64_MAX
```

According to a comment in `<stdint.h>` right above where `UINT64_MAX` is defined:

```c
/* The ISO C99 standard specifies that in C++ implementations these         
   macros should only be defined if explicitly requested.  */
```

I don't think we should force C++ programs to request that these macros be defined.  I'm not a language expert to know what the Right way is to do this.  Here's one suggestion -- is this acceptable?

```c
#define FI_ADDR_UNSPEC     ((uint64_t) ~0)
#define FI_ADDR_NOTAVAIL   ((uin64_t) ~0)
```

***EDIT:*** Added `(uint64_t)` cast.

I'm happy to make a PR with whatever the Right solution is deemed to be.

**Comment by jsquyres** [comment]
@goodell You're good at this kind of stuff -- what's the Right answer here?

**Comment by jsquyres** [comment]
***EDIT:*** Added `(uint64_t)` cast to the original description.

**Comment by goodell** [comment]
Your suggestion is pretty close.  If you want to take this approach then I think you need to make sure it's a large unsigned type first, then cast it (potentially truncating it) to a `uint64_t`.  Something like:

```c++
#define FI_ADDR_UNSPEC     ((uint64_t) ~0ULL)
#define FI_ADDR_NOTAVAIL   ((uint64_t) ~0ULL)
```

But as always, a `printf`/`std::cout<<` test is warranted to be certain.

IIRC, `uint64_t` was only standardized for C++ in C++11, though I suspect it was pretty widely available before that point in all the implementations we are likely to encounter with libfabric.  I think you access them with `#include <cstdint>` and the typedef can technically be optional, though will always be present in practice on modern, mainstream machines.  But someone would need to do the standards archaeology and implementation survey to be sure.

You could do something like this, but I think it might actually be less clear to a reader:

```c++
#if defined(__cplusplus)
#  include <limits>
// for use in this header only
#  define FI_UINT64_MAX__ std::numeric_limits<uint64_t>
#else // !defined(__cplusplus)
#  define FI_UINT64_MAX__ UINT64_MAX
#endif // !defined(__cplusplus)

#define FI_ADDR_UNSPEC     FI_UINT64_MAX__
```

That also doesn't address the potential portability of `uint64_t` among C++ compilers.

**Comment by jsquyres** [comment]
@goodell and I chatted further about this.  Conclusions:

1. We should add an include of `<cstdint>` in `<fabric.h> in the C++ case (because we use `uint64_t` all over the place in the libfabric API)
1. According to https://www.securecoding.cert.org/confluence/display/c/INT17-C.+Define+integer+constants+in+an+implementation-independent+manner and http://stackoverflow.com/questions/809227/is-it-safe-to-use-1-to-set-all-bits-to-true, we should set `FI_ADDR_[UNSPEC|NOTAVAIL]` to -1, and cast it to the right type.  E.g.,

```c
#define FI_ADDR_UNSPEC ((uint64_t) -1)
#define FI_ADDR_NOTAVAIL ((uint64_t) -1)
```

**Comment by jsquyres** [comment]
Arrgh.  g++ v5.1 and v5.2 (the only versions I tested) balk at trying to include `<cstdint>`:

```
$ g++ foo.c
In file included from /usr/local/include/c++/5.1.0/cstdint:35:0,
                 from foo.c:4:
/usr/local/include/c++/5.1.0/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support is currently experimental, and must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
```

Soooo... I'm not going to try to solve that one today.  I'll just change the defines of `FI_ADDR_[UNSPEC|NOTAVAIL]`.
